far/vmenu.cpp

```diff
@@ -1834,15 +1875,69 @@ bool VMenu::ShiftAllItemsShowPos(int Shift)
 {
 	bool NeedRedraw = false;
 
+	const auto MaxLineWidth{ CalculateMaxLineWidth() };
+
 	for (const auto I: std::views::iota(size_t{}, Items.size()))
-		NeedRedraw |= ShiftItemShowPos(static_cast<int>(I), Shift);
+		NeedRedraw |= ShiftItemShowPos(static_cast<int>(I), Shift, MaxLineWidth);
 
 	return NeedRedraw;
 }
 
+struct item_layout
+{
+	std::optional<short> LeftBox;
+	std::optional<short> CheckMark;
+	std::optional<short> LeftHScroll;
+	std::optional<std::pair<short, short>> Text; // Begin, Width
+	std::optional<short> RightHScroll;
+	std::optional<short> SubMenu;
+	std::optional<short> Scrollbar;
+	std::optional<short> RightBox;
+
+	item_layout(const VMenu& Menu, int BoxType)
+	{
+		auto Left{ Menu.m_Where.left };
+		if (NeedBox(BoxType))        LeftBox = Left++;
+		if (NeedCheckMark())         CheckMark = Left++;
+		if (NeedLeftHScroll())       LeftHScroll = Left++;
+
+		auto Right{ Menu.m_Where.right };
+		if (NeedBox(BoxType))        RightBox = Right;
+		if (NeedScrollbar(Menu))     Scrollbar = Right;
+		if (RightBox || Scrollbar)   Right--;
+		if (NeedSubMenu(Menu))       SubMenu = Right--;
+		if (NeedRightHScroll())      RightHScroll = Right--;
+
+		if (Left <= Right)
+			Text = { Left, Right + 1 - Left };
+	}
+
+	[[nodiscard]] static bool NeedBox(int BoxType) noexcept { return BoxType != NO_BOX; }
```

Comment:
Should these static methods be in the struct?
If yes (e.g. to and highlight that they're about layout) it's fine, but it would be better to keep the same case style: `item_layout::GetServiceAreaSize` looks a bit eclectic.

Reply:
Two of them, `NeedSubMenu` and `NeedScrollbar`, need access to `VMenu` privates, as well as `item_layout` constructor. For this purpose, `item_layout` is made a friend. I'll rename them to snake_case.